<h1>Содержание<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Загрузка-данных" data-toc-modified-id="Загрузка-данных-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Загрузка данных</a></span></li><li><span><a href="#Умножение-матриц" data-toc-modified-id="Умножение-матриц-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Умножение матриц</a></span></li><li><span><a href="#Алгоритм-преобразования" data-toc-modified-id="Алгоритм-преобразования-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Алгоритм преобразования</a></span></li><li><span><a href="#Проверка-алгоритма" data-toc-modified-id="Проверка-алгоритма-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Проверка алгоритма</a></span></li><li><span><a href="#Чек-лист-проверки" data-toc-modified-id="Чек-лист-проверки-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Чек-лист проверки</a></span></li></ul></div>

# Защита персональных данных клиентов

Вам нужно защитить данные клиентов страховой компании «Хоть потоп». Разработайте такой метод преобразования данных, чтобы по ним было сложно восстановить персональную информацию. Обоснуйте корректность его работы.

Нужно защитить данные, чтобы при преобразовании качество моделей машинного обучения не ухудшилось. Подбирать наилучшую модель не требуется.

## Загрузка данных

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score

In [2]:
df = pd.read_csv('/datasets/insurance.csv')

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Пол                5000 non-null   int64  
 1   Возраст            5000 non-null   float64
 2   Зарплата           5000 non-null   float64
 3   Члены семьи        5000 non-null   int64  
 4   Страховые выплаты  5000 non-null   int64  
dtypes: float64(2), int64(3)
memory usage: 195.4 KB


Проверим на наличие дупликатов, при наличии удалим их

In [4]:
df.duplicated().sum()

153

In [5]:
df=df.drop_duplicates()
df.shape

(4847, 5)

Посмотрим распределение значений.

In [6]:
df.describe()

Пол      Возраст      Зарплата  Члены семьи  Страховые выплаты
count  4847.000000  4847.000000   4847.000000  4847.000000        4847.000000
mean      0.498453    31.023932  39895.811842     1.203425           0.152259
std       0.500049     8.487995   9972.953985     1.098664           0.468934
min       0.000000    18.000000   5300.000000     0.000000           0.000000
25%       0.000000    24.000000  33200.000000     0.000000           0.000000
50%       0.000000    30.000000  40200.000000     1.000000           0.000000
75%       1.000000    37.000000  46600.000000     2.000000           0.000000
max       1.000000    65.000000  79000.000000     6.000000           5.000000

Выбросов не замечено.

Преобразуем колонки в целое число

In [7]:
df['Возраст'] = df['Возраст'].astype('int')
df['Зарплата'] = df['Зарплата'].astype('int')


Проверим уникальные значения для оценки данных на наличие аномалий

In [8]:
print(df['Пол'].unique())
print(df['Возраст'].unique())
print(df['Члены семьи'].unique())
print(df['Страховые выплаты'].unique())

[1 0]
[41 46 29 21 28 43 39 25 36 32 38 23 40 34 26 42 27 33 47 30 19 31 22 20
 24 18 37 48 45 44 52 49 35 56 65 55 57 54 50 53 51 58 59 60 61 62]
[1 0 2 4 3 5 6]
[0 1 2 3 5 4]


Аномалии не выявлены

Итого получили данные с 5 столбцами и 4847 строками 

In [9]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4847 entries, 0 to 4999
Data columns (total 5 columns):
 #   Column             Non-Null Count  Dtype
---  ------             --------------  -----
 0   Пол                4847 non-null   int64
 1   Возраст            4847 non-null   int64
 2   Зарплата           4847 non-null   int64
 3   Члены семьи        4847 non-null   int64
 4   Страховые выплаты  4847 non-null   int64
dtypes: int64(5)
memory usage: 227.2 KB


## Умножение матриц

В этом задании вы можете записывать формулы в *Jupyter Notebook.*

Чтобы записать формулу внутри текста, окружите её символами доллара \\$; если снаружи —  двойными символами \\$\\$. Эти формулы записываются на языке вёрстки *LaTeX.* 

Для примера мы записали формулы линейной регрессии. Можете их скопировать и отредактировать, чтобы решить задачу.

Работать в *LaTeX* необязательно.

Обозначения:

- $X$ — матрица признаков (нулевой столбец состоит из единиц)

- $y$ — вектор целевого признака

- $P$ — матрица, на которую умножаются признаки

- $w$ — вектор весов линейной регрессии (нулевой элемент равен сдвигу)

Предсказания:

$$
a = Xw
$$

Задача обучения:

$$
w = \arg\min_w MSE(Xw, y)
$$

Формула обучения:

$$
w = (X^T X)^{-1} X^T y
$$

Признаки умножают на обратимую матрицу. 

Изменится ли качество линейной регрессии? (Её можно обучить заново.)

**Ответ:** 

Не изменится. Укажите, как связаны параметры линейной регрессии в исходной задаче и в преобразованной.

**Обоснование:** 

$$
a = Xw
$$
Преобразуем матрицу
$$
X1 = XP
$$

Подставим значение Х1 в формулы
$$
a1 = XPw1
$$

$$
w = \arg\min_w MSE(Xw, y)
$$

$$
w1 = (X1^T X1)^{-1} X1^T y
$$


$$
w1 = ((XP)^T (XP))^{-1} (XP)^T y 
$$

Используем следующие свойства матриц для преобразования:

$$
(AB)^T = B^T A^T
$$
$$
AA^{-1} = E
$$
$$
AE = A
$$
$$
(A)^{-1} = B^{-1} A^{-1}
$$

Преобразуем а1
$$
a1=XPw1=XP((XP)^T (XP))^{-1} (XP)^T y = XP (XP)^{-1} ((XP)^T)^{-1} (XP)^T y = XPP^{-1} X^{-1}((XP)^T)^{-1} (XP)^T y = XX^{-1}(P^TX^T)^{-1}P^TX^T y=XX^{-1}(X^T)^{-1} (P^T)^{-1} P^T X^T y = XX^{-1}(X^T)X^Ty=X(X^TX)^{-1} X^T y = Xw = a
$$


**Ответ:** 

Качество линейной регрессии не изменится в случае, если матрица Р невырожденная, т.е. детерминант не равен 0



$$
a = Xw = XEw = XPP^{-1}w = (XP)P^{-1}w = (XP)w'
$$

$$
w' = ((XP)^T XP)^{-1} (XP)^T y
$$
$$
w' = (P^T (X^T X) P)^{-1} (XP)^T y
$$
$$
w' = ((X^T X) P)^{-1} (P^T)^{-1} P^T X^T y
$$
$$
w' = P^{-1}(X^T X)^{-1} X^T y
$$

$$
w' = P^{-1}w 
$$


    
Если есть время и желание, то можно воту этот маткриал просмотреть:<br>
Обзоры по применению криптографии в машинном обучении: https://arxiv.org/pdf/1902.04109.pdf, https://www.researchgate.net/publication/342746786_Some_Applications_of_Machine_Learning_in_Cryptography
Подробнее про аналогичный алгоритм: https://www.cs.cmu.edu/~rjhall/JOS_revised_May_31a.pdf<br><br>


## Алгоритм преобразования

**Алгоритм**

Испоьзуем доказательство выше

1. Построим случайную квадратную  матрицу Р, размерностью равную колличеству признаков
2. Проверим значение детерминанта, т.е. проверка матрицы на обратимость 

3. Получим матрицу Х1 = РХ

4. Построи модель, проверим значения на данных до преобразования и после. 


Создадим функцию преобразования данных

In [10]:
def transformation(matrix):
    n = matrix.shape[1]
    np.random.seed(12345)
    random_matrix = np.random.randint(1, 10, (n,n))
    det = np.linalg.det(random_matrix)
    if det == 0:
        np.random.seed(123456)
        random_matrix = np.random.randint(1, 10, (n,n))
        det = np.linalg.det(random_matrix)
        if det == 0:
            print('Error!')
    transform_matrix = matrix @ random_matrix
    return transform_matrix, random_matrix

In [11]:
df_matrix = np.array(df)
df_matrix

array([[    1,    41, 49600,     1,     0],
       [    0,    46, 38000,     1,     1],
       [    0,    29, 21000,     0,     0],
       ...,
       [    0,    20, 33900,     2,     0],
       [    1,    22, 32700,     3,     0],
       [    1,    28, 40600,     1,     0]])

Преобразем данные, используя написанную функцию

In [12]:
df_transform_matrix, random_matrix = transformation(df_matrix)
df_transform_matrix

array([[347333,  49696, 149097,  99295, 149143],
       [266153,  38102, 114338,  76105, 114378],
       [147087,  21058,  63203,  42058,  63232],
       ...,
       [237374,  33956, 101856,  67856, 101878],
       [228990,  32774,  98280,  65473,  98309],
       [284294,  40670, 122006,  81269, 122039]])

Заметим, что получить исходные данные из полученной матрицы, без random_matrix затруднительно

In [13]:
random_matrix

array([[3, 6, 2, 5, 6],
       [3, 2, 7, 2, 8],
       [7, 1, 3, 2, 3],
       [7, 8, 8, 8, 9],
       [8, 2, 8, 5, 1]])

Проверим детерминант 

In [14]:
np.linalg.det(random_matrix)

457.9999999999998

**Обоснование**

Теоретические доказательства приведены в п. 2

Посчитаем вектор весов и вектор предсказаний до преобразования и после

In [15]:
features = np.array(df.drop('Страховые выплаты', axis=1))
target = np.array(df['Страховые выплаты'])


In [16]:
w = np.linalg.inv(np.dot(features.T,features)).dot(features.T).dot(target)
a = features.dot(w)

In [17]:
features_tr, random_matrix = transformation(features)

In [18]:
w1 = np.linalg.inv(np.dot(features_tr.T,features_tr)).dot(features_tr.T).dot(target)
a1 = features_tr.dot(w1)

Проверим разницу между векторами предсказаний до преобразования и после

In [19]:
np.allclose(a,a1,atol=1e-7)

False

## Проверка алгоритма

Построим модель, обучим на данных до преобразования и после преобразования

In [20]:
features_train, features_test, target_train, target_test = train_test_split(
    features, target, test_size=0.25, random_state=12345)

In [21]:
model = LinearRegression()
model.fit(features_train, target_train)
R2_LR_df = r2_score(target_test, model.predict(features_test))
print("R2 =", R2_LR_df)

R2 = 0.42307727615837565


In [22]:
features_tr_train, features_tr_test, target_train, target_test = train_test_split(
    features_tr, target, test_size=0.25, random_state=12345)

In [23]:
model = LinearRegression()
model.fit(features_tr_train, target_train)
R2_LR_df_tr = r2_score(target_test, model.predict(features_tr_test))
print("R2 =", R2_LR_df_tr)

R2 = 0.42307727615839963


In [24]:
result = pd.DataFrame([R2_LR_df,R2_LR_df_tr], columns=['R2'], index=['Линейная регрессия','Линейная регрессия на преобразованных признаках'])
result

R2
Линейная регрессия                               0.423077
Линейная регрессия на преобразованных признаках  0.423077

Выводы: Качество линейной регрессии не изменится после преобразования данных, путем путем умножения на обратимую матрицу. 

## Чек-лист проверки

Поставьте 'x' в выполненных пунктах. Далее нажмите Shift+Enter.

- [x]  Jupyter Notebook открыт
- [x]  Весь код выполняется без ошибок
- [x]  Ячейки с кодом расположены в порядке исполнения
- [x]  Выполнен шаг 1: данные загружены
- [x]  Выполнен шаг 2: получен ответ на вопрос об умножении матриц
    - [x]  Указан правильный вариант ответа
    - [x]  Вариант обоснован
- [x]  Выполнен шаг 3: предложен алгоритм преобразования
    - [x]  Алгоритм описан
    - [x]  Алгоритм обоснован
- [x]  Выполнен шаг 4: алгоритм проверен
    - [x]  Алгоритм реализован
    - [x]  Проведено сравнение качества моделей до и после преобразования